### ISSUE TITLE
Enabling proofs leads to crash in statistics module (Bugzilla #688)

### ISSUE THREAD

**Issue by mpreiner** [CVC4/CVC4#907]
**Imported from Bugzilla**
Reporter: *Kshitij*
Status: *REOPENED*
Severity: *normal*
Assigned to: *Kshitij*
Component: *{Internals} Utilities*
Milestone: *1.6*
Version: *master*
Platform: *PC*
OS: *Linux*

Original attachment names and IDs:
- _[gdb-session-1.txt](http://cvc4.cs.stanford.edu/bugzilla_attachments/331_gdb-session-1.txt)_ (gdb session where statistics registry is NULL, ID 331)
- _[gdb-session-2.txt](http://cvc4.cs.stanford.edu/bugzilla_attachments/332_gdb-session-2.txt)_ (gdb session where d_stats in SmtEngine is corrupted, ID 332)
- _[patch](http://cvc4.cs.stanford.edu/bugzilla_attachments/333_patch)_ (patch mentioned in comment 4, ID 333)

On 2015-10-24 21:22:27 -0700, Kshitij wrote:
> Here's the script to reproduce:
> 
> \=\=\=\=\=\=
> \#!/bin/bash
> set \-e
> rm \-fr builds
> rm \-fr INSTALLATIONdebug
> ./autogen.sh
> ./configure personal
> make \-j7 install
> make \-j7 install\-examples
> cd INSTALLATIONdebug
> libtool \-\-mode\=link g\+\+ \-g \-Iinclude \-Iinclude/cvc4 \-Llib share/doc/cvc4/examples/simple\_vc\_cxx.cpp \-o cvc4\-example \-lcvc4 \-lcln \-lgmp \-lgmpxx
> ./cvc4\-example || exit 1
> \=\=\=\=\=
> Exit 0: good
> Exit 1: bad
> 
> personal.conf (paths should be modified)
> \=\=\=\=\=
> debug
> \-\-with\-antlr\-dir\=/home/kshitij/cvc/CVC4.CVC4/antlr\-3.4
> ANTLR\=/home/kshitij/cvc/CVC4.CVC4/antlr\-3.4/bin/antlr3
> \-\-with\-cxxtest\-dir\=/home/acsys\-user/software/cxxtest
> \-\-prefix\=/home/kshitij/data/CVC4.proofsenable/INSTALLATIONdebug
> \-\-enable\-proof
> \=\=\=\=\=
> 
> \-\-Kshitij

On 2015-10-24 21:42:18 -0700, Kshitij wrote:
> More information:
> 
> The areas I have investigated are
>   src/smt/smt\_engine.cpp
> and
>   src/util/statistics\_registry.cpp
> 
> In the intialization of SmtEngine::SmtEngine(ExprManager\* em), we have
> 
>    724    d\_statisticsRegistry \= new StatisticsRegistry();
>    725    d\_stats \= new SmtEngineStatistics();
> 
> In some runs, d\_stats pointer gets corrupted. In the intialization function the value will be fine.
> 
> But breaking at the crash point, the value of d\_stats will be
>    (CVC4::smt::SmtEngineStatistics \*) 0x100000001
> This is at line number 805 in smt\_engine.cpp
> 
> That is if you look at value of d\_stats at initialziation it will be a usual pointer, but at crashing point if you go up in gdb and look at value of d\_stats it is garbage.
> 
> Then of course, things crash when trying to add stats.
> 
> In other runs of gdb, and the value of d\_statisticsRegistry is NULL even though that new appears to be running. Not sure what is going on.
> 
> Also, if I look at:
>   src/util/statistics\_registry.cpp
> 
> there is lot of code in "ifdef"s. Do the build rules for proofs somehow affect some of these variables?
> 
> Kshitij

On 2015-10-24 21:46:53 -0700, Kshitij wrote:
> Created attachment 331
> gdb session where statistics registry is NULL

On 2015-10-24 21:47:16 -0700, Kshitij wrote:
> Created attachment 332
> gdb session where d\_stats in SmtEngine is corrupted

On 2015-10-25 02:21:23 -0700, Tim King wrote:
> Including my email to cvc4\-devel.
> \-\-
> Great catch on this being proofs Kshitij!
> 
> tl;dr No CVC4\_PROOF \#ifdefs in public headers! We need an enforcement mechanism to make this harder in the future.
> 
> The test that is failing is PerformInstallCheck() in \~acsys\-user/tmp/build\_and\_test.sh . What is interesting about this test is that it compiles with different flags than CVC4. This is just like what a user would do! Here is a simplified view of what the test does:
>   ./configure ... \-\-prefix\=INSTALLATION
>   make install install\-examples
>   cd INSTALLATION
>   libtool \-\-mode\=link g\+\+ \-Iinclude \-Iinclude/cvc4 \-Llib share/doc/cvc4/examples/simple\_vc\_cxx.cpp \-o cvc4\-example \-lcvc4 \-lcln \-lgmp \-lgmpxx 
>  ./cvc4\-example
> That last line is the test that segfaults.  You can see that if you add \-DCVC4\_PROOF, recompile and rerun the test succeeds.
> 
> There is an \#ifdef that depends on CVC4\_PROOF around in src/smt/smt\_engine.h . Removing this as follows also makes the test pass.
> diff \-\-git a/src/smt/smt\_engine.h b/src/smt/smt\_engine.h
> index b4c293d..b3ff8d4 100644
> \-\-\- a/src/smt/smt\_engine.h
> \+\+\+ b/src/smt/smt\_engine.h
> _@_@ \-196,9 \+196,7 _@_@ class CVC4\_PUBLIC SmtEngine {
>     \*A vector of command definitions to be imported in the new
>     \*SmtEngine when checking unsat\-cores.
>     \*/
> \-\#ifdef CVC4\_PROOF
>    std::vector<Command\*> d\_defineCommands;
> \-\#endif
>    /\*\*
>     \* The logic we're in.
>     \*/
> 
> Going back to the original code we can confirm that sizeof(SmtEngine) is different internally and while compiling simple\_vc\_cxx.cpp. I have attached a patch that shows the behavior below.
> $ ./cvc4\-example 
> Size of SmtEngine from main: 384
> Size of SmtEngine from SmtEngine::SmtEngine: 408
> Checking validity of formula ((x > 0) AND (y > 0)) \=> (((2 \* x) \+ y) >\= 3) with CVC4.
> CVC4 should report VALID.
> smt scope: 0x7fffcf6e2c10
> Segmentation fault (core dumped)
>  
> 
> The segfault that we see is caused by memory corruption of the d\_statisticsRegistry pointer in the SmtEngine. This looks consistent with two parts of the compiler seeing different sizes for the same class. This was really annoying to pin down. (And would have been a massive pain without Kshitij's initial observation that it was proofs.)
> 
> The only safe way to avoid problems like in the future is to figure out a way to enforce that none of the compile time flags we use appears in a public header. Any thoughts for how to achieve this?
> 
> \-\-Tim

On 2015-10-25 02:22:03 -0700, Tim King wrote:
> Created attachment 333
> patch mentioned in comment 4

On 2015-10-26 14:18:00 -0700, Kshitij wrote:
> Thanks Tim, good detective work. Let's discuss in the meeting this week for a robust solution to avoid this in the future (and also detect if there other ifdefs we haven't detected in our usual compilation cycles).
> 
> \-\-Kshitij

On 2015-11-04 02:09:34 -0800, Tim King wrote:
> Fixed in commit 0f66dd16f35eac64149919f0f4048b422345c5eb.

On 2015-11-04 03:12:58 -0800, Kshitij wrote:
> Leaving this open till I add the checks to nightly builds. You could unsubscribe from it Tim :\-) thanks for your help! \-\-Kshitij

On 2016-12-07 17:58:20 -0800, Clark Barrett wrote:
> Kshitij \- is this bug still relevant?

**Comment by yoni206** [comment]
Andres and I have tested this by installing cvc4, compiling and running Kshitij's example, and the problem does not reproduce.
